orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -2178,6 +2178,25 @@ Return true if all elements are true and false otherwise.
           propagateShapeFromInputToOutput(ctx, i + 1, i);
         }
       });
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(TopKGrad)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetSupportLevel(OpSchema::SupportType::EXPERIMENTAL)
+      .SetDoc("TopKGrad")
+      .AllowUncheckedAttributes()
```

Comment:
need the axis attribute? 

The Indices output is index on a specific axis

Reply:
Do you know if it would be possible to use `GatherGrad` (or maybe `GatherElementsGrad`... I'm not sure) to implement this gradient, instead of adding a new kernel?

I did some tests with PyTorch and found that the following equivalence seemed to hold:

```
x = torch.rand([2, 3, 4, 5])
topk_values, topk_indices = x.topk(3, dim=2)
gather_values = x.gather(topk_indices, dim=2)
print((topk_values == gather_values).all())  # prints "tensor(True)"
```

I think this implies that we could use the gradient of gather to implement the gradient of top-k, but there could be a corner case that I haven't seen. It would be nice to reuse GatherGrad though, because there's been a lot of performance engineering on it recently.

Reply:
Thanks @mrry ! You are right that GatherElementsGrad can be reused here according to the Op definition. But current GatherElementsGrad doesn't have CPU kernel implementation, then my CPU code can be changed to GatherElementsGrad. :-) Then both GatherElementsGrad and TopKGrad will have both CPU and CUDA. Another bad news is when I use GatherElementsGrad for TopKGrad, it failed one of my UT using CUDA (my CPU implementation works OK), I then added a new UT with same data size and attributes for GatherElementsGrad and it also failed. It means current CUDA implementation has bug somewhere when axis attribute is not default. I will investigate and fix it.

Reply:
Just pushed a new version to use Scatter CPU/CUDA impl for both GatherElementsGrad and TopKGrad.